five/intid/keyreference.py

```diff
@@ -154,13 +155,18 @@ def __call__(self):
         return self.wrapped_object
 
     def __hash__(self):
-        # XXX Maybe we should consider to use also other fields for the hash
         return hash((self.dbname,
                      self.object._p_oid,
+                     self.path
                      ))
 
     def __cmp__(self, other):
         # XXX This makes no sense on Python 3
         if self.key_type_id == other.key_type_id:
-            return cmp((self.dbname, self.oid), (other.dbname, other.oid))
+            return cmp((self.dbname, self.oid, self.path), (other.dbname, other.oid, other.path))
         return cmp(self.key_type_id, other.key_type_id)
+
+    def _get_cmp_keys(self, other):
+        if self.key_type_id == other.key_type_id:
+            return (self.dbname, self.oid, self.path), (other.dbname, other.oid, other.path)
```

Comment:
I see that the original method is way more complex and with code that should help conflict resolution: 
https://github.com/zopefoundation/zope.keyreference/blob/master/src/zope/keyreference/persistent.py#L61

We could probably do something like:
```python
def _get_cmp_keys(self, other):
    if self.key_type_id != other.key_type_id:
        return self.key_type_id, other.key_type_id
    keys = super(KeyReferenceToPersistent, self)._get_cmp_keys(self, other)
    # The original method does not take in to account the path
    return keys[0] + (self.path,), keys[1] + (self.path,)
```
Which is ugly but could do its job.
What do you think @mauritsvanrees? Am I missing something?

Reply:
Done, and I did the same for the `__hash__` method where upstream has similar code for conflict resolution.

But I am investigating if this PR is still needed. Hang on.